Feature/home continue watching

saraashraf11 opened the pull request:
# Pull Request Template

## Description

This pull request introduces the complete implementation of the Continue Watching feature.
It enables users to view and resume movies or shows they haven't finished, improving the app’s user experience and engagement.

---

## Changes Made

- Created ContinueWatchingScreen and ContinueWatchingContent using Jetpack Compose
- Implemented ContinueWatchingMoviesGrid to display partially watched items
- Handled UI states (loading, error, empty, and success)
- Implemented ContinueWatchingViewModel with a corresponding UiState and Effect
- Added ContinueWatchingInteractionListener for user actions like retry and navigation
- Collected state and effects using collectAsStateWithLifecycle and LaunchedEffect
- Created GetContinueWatchingMoviesUseCase for business logic and AddWatchHistoryUseCase
- Defined necessary interfaces in the domain repository
- Added local data model WatchHistoryDto with appropriate DAO queries
- Implemented LocalContinueWatchingDataSource and LocalContinueWatchingDataSourceImpl
- Created and implemented ContinueWatchingRepositoryImpl
- Registered all new components in the DI module (ViewModel, Repository, Data Source, Use Case)

---

## Screenshots (if applicable)

▶️ [Click here to watch the demo video](https://drive.google.com/file/d/1Vufy1DUxzn9DR4VjvliAbIprIpgM4k0H/view?usp=sharing)


---

## Checklist
Please ensure the following tasks are completed:

- [x] My code follows the code style of this project
- [x] Changes have been tested manually and verified.
- [x] PR includes at most one single feature.

saraashraf11 committed:
feature/ add di modules

saraashraf11 committed:
feature/ add AddWatchHistoryUseCase and GetContinueWatchingMoviesUseCase

saraashraf11 committed:
feature/ add WatchHistoryRepository and it's implementation

saraashraf11 committed:
feature/ add WatchHistoryLocalDataSource and it's implementation

saraashraf11 committed:
feature/ add WatchHistoryDto

saraashraf11 committed:
feature/ add WatchHistoryDao and database

saraashraf11 committed:
feature/ add WatchHistoryMapper

saraashraf11 committed:
feature/ add continueWatchingScreenRoute

saraashraf11 committed:
feature/ add ContinueWatching effect and interaction listener

saraashraf11 committed:
feature/ add ContinueWatchingUiState

saraashraf11 committed:
feature/ add ContinueWatchingViewModel

saraashraf11 committed:
feature/ add Home effect and interaction listener

saraashraf11 committed:
feature/ add ContinueWatchingScreen

saraashraf11 committed:
feature/ update HomeScreen

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into feature/home-continue-watching

# Conflicts:
#	app/src/main/java/com/amsterdam/aflami/di/localDataSourceModule.kt
#	app/src/main/java/com/amsterdam/aflami/di/repositoryModule.kt
#	app/src/main/java/com/amsterdam/aflami/di/useCaceModule.kt
#	app/src/main/java/com/amsterdam/aflami/di/viewModelModule.kt
#	domain/src/main/java/com/example/domain/useCase/GetMovieDetailsUseCase.kt
#	localDatasource/src/main/java/com/example/localdatasource/roomDataBase/AflamiDatabase.kt
#	ui/src/main/java/com/example/ui/navigation/NavGraph.kt
#	ui/src/main/java/com/example/ui/navigation/Route.kt

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into feature/home-continue-watching

# Conflicts:
#	viewModel/src/main/java/com/example/viewmodel/home/HomeViewModel.kt

Commit:
Refactor: Improve "Continue Watching" feature

This commit enhances the "Continue Watching" functionality by:

- Modifying `WatchHistoryDao` to use `@Upsert` for `addToWatchHistory` and to return a `Flow` for `getContinueWatching`.
- Updating `WatchHistoryLocalDataSource` and its implementation to reflect the Dao changes.
- Adjusting `WatchHistoryRepository` and its implementation to work with `Flow`.
- Updating `GetContinueWatchingMoviesUseCase` to return a `Flow`.
- Modifying `AddWatchHistoryUseCase` to use `Clock.System.now()` for `lastWatchedTime`.
- Changing `WatchHistory.lastWatchedTime` to `Instant`.
- Changing `WatchHistoryDto.lastWatchedTime` to `Instant` and making `movieId` the primary key.
- Refactoring `HomeViewModel` to observe continue watching movies using a `Flow`.
- Refactoring `ContinueWatchingViewModel` to load movies using `firstOrNull()` from the `Flow`.
- Updating `GetHomeScreenDataUseCase` to use `firstOrNull()` when fetching continue watching movies.
- Ensuring `MovieRepositoryImpl` caches watched movies locally.
- Adding `insertMovie` function to `MovieLocalDataSource` and `MovieLocalDataSourceImpl`.
- Removing unused import in `useCaceModule.kt`.
- Minor UI adjustments in `ContinueWatchingScreen.kt` and `HomeScreen.kt`.
- Fixing package name in `ResetPasswordWebView.kt`.
- Removing unused import in `ResetPasswordScreenRout.kt`.
- Fixing package name in `ResetPasswordScreen.kt`.
- Removing unused import in `AflamiDatabase.kt`.
- Correcting year of release in `ContinueWatchingUiStateMapper.kt`.
- Removing completion handler from `MovieDetailsViewModel.kt` error handling.

Belal-Alkhatib commented on repository/src/main/java/com/example/repository/repository/WatchHistoryRepositoryImpl.kt:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    override suspend fun getContinueWatchingMovies(): Flow<List<Movie>> {
+        return flow {
+            watchHistoryLocalDataSource.getContinueWatching().collect{
+                emit(movieLocalMapper.toEntityList(it))
+            }
+        }
```
you can use map instead of wrapping the flow in a flow { collect { emit() } } block

somthing like:
```
override suspend fun getContinueWatchingMovies(): Flow<List<Movie>> {
    return watchHistoryLocalDataSource.getContinueWatching()
        .map { movieLocalMapper.toEntityList(it) }
}
```

Belal-Alkhatib commented on ui/src/main/java/com/example/ui/screens/continueWatching/ContinueWatchingScreen.kt:
```diff
@@ -0,0 +1,117 @@
+            it?.let {
+                when (it) {
+                    is ContinueWatchingEffect.NavigateToMovieDetailsScreen -> {
+                        navController.navigate(Route.MovieDetails(it.movieId))
```
please use our safeNavigate extension function instead of navigate to avoid potential duplicate destination issues in the back stack

Commit:
refactor: simplify Flow emission in WatchHistoryRepositoryImpl

The `getContinueWatchingMovies` function in `WatchHistoryRepositoryImpl` has been refactored to use the `map` operator for transforming the Flow of watch history data into a Flow of `Movie` entities. This simplifies the previous implementation that used `flow` builder and explicit `collect` and `emit`.

Additionally, the `ContinueWatchingScreen` now uses `safeNavigate` for navigating to movie details, enhancing navigation safety.
